### PR TITLE
Colormix

### PR DESCRIPTION
**Pull request by samthiele**
Added tool to build RGB colours from multiple scalar fields (_Edit->Colors->From Scalar fields_).

### CODE REVIEW

File: qCC/ccColorFromScalarDlg.cpp [hunk 1]
```diff
@@ -0,0 +1,476 @@
...
+//#                                                                        #
+//#          COPYRIGHT: CloudCompare project                               #
+//#                                                                        #
+//##########################################################################
+
+#include "ccColorFromScalarDlg.h"
+
+//local
+#include "ccHistogramWindow.h"
+//#include "ccGenericPointCloud.h"
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Remove

File: qCC/ccColorFromScalarDlg.cpp [hunk 2]
```diff
@@ -0,0 +1,476 @@
...
+#include <ccGenericGLDisplay.h>
+#include <ccPointCloud.h>
+#include <ccHObjectCaster.h>
+#include <ccColorScale.h>
+
+//Qt
+#include <QPushButton>
+
+//system
+#include <string.h>
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Should be cstring and cassert, not the older C headers.

File: qCC/ccColorFromScalarDlg.cpp [hunk 3]
```diff
@@ -0,0 +1,476 @@
...
+#include <assert.h>
+
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: All includes should be alphabetized within their groupings.

File: qCC/ccColorFromScalarDlg.cpp [hunk 4]
```diff
@@ -0,0 +1,476 @@
...
+//persistent parameters
+static int s_inputLevels[2] = {0,255};
+static int s_outputLevels[2] = {0,255};
+static bool s_outputLevelsEnabled = false;
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Should these be included in the class instead of floating around here?

File: qCC/ccColorFromScalarDlg.cpp [hunk 5]
```diff
@@ -0,0 +1,476 @@
...
+	}
+	updateColormaps();
+
+	//update histograms
+	updateChannel(0);
+	updateChannel(1);
+	updateChannel(2);
+	updateChannel(3);
+	m_cloud->getCurrentDisplayedScalarField()->setColorScale(m_colors[3]); //set grey colour ramp to start with
+
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Remove extra blank line

File: qCC/ccColorFromScalarDlg.cpp [hunk 6]
```diff
@@ -0,0 +1,476 @@
...
+	//update histograms
+	updateChannel(0);
+	updateChannel(1);
+	updateChannel(2);
+	updateChannel(3);
+}
+
+void ccColorFromScalarDlg::updateHistogram(int n)
+{
+
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Remove extra blank line

File: qCC/ccColorFromScalarDlg.cpp [hunk 7]
```diff
@@ -0,0 +1,476 @@
...
+		m_boxes_max[n]->setMaximum(255.0);
+		m_boxes_min[n]->setValue(200.0);
+		m_boxes_max[n]->setValue(200.0);
+
+		//edge case for HSV values (0 - 360)
+		if (n == 0 && toggleHSV->isChecked())
+		{
+			m_boxes_min[n]->setMaximum(360);
+		}
+
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Remove extra blank line

File: qCC/ccColorFromScalarDlg.cpp [hunk 8]
```diff
@@ -0,0 +1,476 @@
...
+	}
+}
+
+void ccColorFromScalarDlg::updateChannel(int n)
+{
+	ccScalarField* sf = static_cast<ccScalarField*>(m_cloud->getScalarField(m_combos[n]->currentIndex()));
+	m_scalars[n] = sf;
+	if (sf)
+	{
+
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Remove extra blank line

File: qCC/ccColorFromScalarDlg.cpp [hunk 9]
```diff
@@ -0,0 +1,476 @@
...
+		m_boxes_max[n]->setMinimum(m_minSat[n]);
+		m_boxes_max[n]->setMaximum(m_maxSat[n]);
+		m_boxes_max[n]->setCorrectionMode(QAbstractSpinBox::CorrectionMode::CorrectToNearestValue);
+		updateHistogram(n);
+
+		//set default stretch (n.b. this is a cheap hack to avoid calculating percentiles [by assuming uniform data distribution])
+		float range = m_maxSat[n] - m_minSat[n];
+		m_histograms[n]->setMinSatValue(m_minSat[n] + 0.1 * range);
+		m_histograms[n]->setMaxSatValue(m_maxSat[n] - 0.1 * range);
+
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Remove extra blank line

File: qCC/ccColorFromScalarDlg.cpp [hunk 10]
```diff
@@ -0,0 +1,476 @@
...
+	}
+}
+
+void ccColorFromScalarDlg::maxChanged(int n, double val, bool slider)
+{
+	if (val >= m_minSat[n]) //valid value?
+	{
+		m_scalars[n]->setColorScale(m_colors[n]);
+		m_maxSat[n] = val;
+		//m_boxes_max[n]->setValue(val-0.1);
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Remove

File: qCC/ccColorFromScalarDlg.cpp [hunk 11]
```diff
@@ -0,0 +1,476 @@
...
+		{
+			m_boxes_max[n]->setValue(val);
+		}
+		else //change was made with text box
+		{
+			m_histograms[n]->setMaxSatValue(val);
+		}
+	}
+}
+
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Remove extra blank line

File: qCC/ccColorFromScalarDlg.cpp [hunk 12]
```diff
@@ -0,0 +1,476 @@
...
+
+void ccColorFromScalarDlg::onApply()
+{
+
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Remove extra blank line

File: qCC/ccColorFromScalarDlg.cpp [hunk 13]
```diff
@@ -0,0 +1,476 @@
...
+			{
+				if (fixed[i]) //fixed value
+				{
+					col[i] = int(m_boxes_min[i]->value());
+				}
+				else //map from scalar
+				{
+					col[i] = int(255.0 * (m_scalars[i]->getValue(p) - m_minSat[i]) / (m_maxSat[i] - m_minSat[i]));
+				}
+
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Remove extra blank line

File: qCC/ccColorFromScalarDlg.cpp [hunk 14]
```diff
@@ -0,0 +1,476 @@
...
+	{
+		int col[4];
+		for (unsigned p = 0; p < m_cloud->size(); p++)
+		{
+			//get col
+			for (unsigned i = 0; i < 4; i++)
+			{
+				if (fixed[i]) //fixed value
+				{
+					col[i] = int(m_boxes_min[i]->value());
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: static_cast&lt;int&gt;(...)

File: qCC/ccColorFromScalarDlg.cpp [hunk 15]
```diff
@@ -0,0 +1,476 @@
...
+				}
+				else //map from scalar
+				{
+					col[i] = int(255.0 * (m_scalars[i]->getValue(p) - m_minSat[i]) / (m_maxSat[i] - m_minSat[i]));
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: static_cast&lt;int&gt;(...)

File: qCC/ccColorFromScalarDlg.cpp [hunk 16]
```diff
@@ -0,0 +1,476 @@
...
+			} else
+			{
+				hue = i;
+			}
+
+			//calculate colour
+			col = ccColor::Convert::hsv2rgb(hue, 1.0f, 1.0f);
+
+			//add stop
+			m_colors[0]->insert(ccColorScaleElement( double(i) / 360.0, QColor( int(col.r), int(col.g), int(col.b), 255)));
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: static_cast&lt;double&gt;(...)
static_cast&lt;int&gt;(...)

File: qCC/ccColorFromScalarDlg.cpp [hunk 17]
```diff
@@ -0,0 +1,476 @@
...
+
+			m_colors[i]->update();
+		}
+
+		//overwrite first colourmap with hue rainbow
+		int hue;
+		m_colors[0]->clear();
+
+		ccColor::Rgb col = ccColor::Convert::hsv2rgb(360.0f, 1.0f, 1.0f);
+		m_colors[0]->insert(ccColorScaleElement(1.0, QColor(int(col.r), int(col.g), int(col.b), 255)));
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: static_cast&lt;int&gt;(...)

File: qCC/ccColorFromScalarDlg.cpp [hunk 18]
```diff
@@ -0,0 +1,476 @@
...
+				//flip?
+				if (reversed[i] && !fixed[i])
+				{
+					col[i] = 1.0f - col[i];
+				}
+			}
+
+			//calculate and set colour
+			ccColor::Rgb rgb = ccColor::Convert::hsv2rgb(col[0] * 360.0f, col[1], col[2]);
+			m_cloud->setPointColor(p, ccColor::FromQColor(QColor(int(rgb.r), int(rgb.g), int(rgb.b), int(col[3] * 255))));
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: static_cast&lt;int&gt;(...)

Is there not a cleaner way to do this conversion? Why are we converting to QColor and then back?

File: qCC/ccColorFromScalarDlg.cpp [hunk 19]
```diff
@@ -0,0 +1,476 @@
...
+		}
+
+	}
+		m_cloud->colorsHaveChanged();
+		m_cloud->showSF(false);
+		m_cloud->showColors(true);
+		m_cloud->redrawDisplay();
+		m_cloud->prepareDisplayForRefresh();
+		m_cloud->refreshDisplay();
+
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: Remove extra blank line

File: qCC/ccColorFromScalarDlg.h [hunk 20]
```diff
@@ -0,0 +1,111 @@
...
+
+#include <ui_colorFromScalarDlg.h>
+#include <ccColorScale.h>
+#include <ccScalarField.h>
+#include <ccHistogramWindow.h>
+
+class ccPointCloud;
+
+//! Dialog to change the color levels
+class ccColorFromScalarDlg : public QDialog, public Ui::ColorFromScalarDialog
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.h]: UI classes shouldn't be done with inheritance.

They should be [pointer members](https://doc.qt.io/qt-5.9/designer-using-a-ui-file.html#using-a-pointer-member-variable).

File: qCC/ccColorFromScalarDlg.h [hunk 21]
```diff
@@ -0,0 +1,111 @@
...
+	Q_OBJECT
+
+public:
+
+	//! Default constructor
+	ccColorFromScalarDlg(QWidget* parent, ccPointCloud* pointCloud);
+
+	//update and redraw histograms
+	void updateHistogram(int);
+protected slots:
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.h]: Add blank line before

File: qCC/ccColorFromScalarDlg.h [hunk 22]
```diff
@@ -0,0 +1,111 @@
...
+	//saturation values
+	double m_minSat[4];
+	double m_maxSat[4];
+
+	//! Associated colour scales
+	ccColorScale::Shared m_colors[4];
+
+	//! Associated point cloud (color source)
+	ccPointCloud* m_cloud;
+
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.h]: Remove blank lines

File: qCC/mainwindow.cpp [hunk 23]
```diff
@@ -117,6 +117,7 @@
 #include "ccVolumeCalcTool.h"
 #include "ccWaveformDialog.h"
 #include "ccEntitySelectionDlg.h"
+#include "ccColorFromScalarDlg.h"
```
Comment (asmaloney) [inline qCC/mainwindow.cpp]: These should be sorted

File: qCC/mainwindow.cpp [hunk 24]
```diff
@@ -506,10 +507,12 @@ void MainWindow::connectActions()
 	connect(m_UI->actionRGBToGreyScale,				&QAction::triggered, this, &MainWindow::doActionRGBToGreyScale);
 	connect(m_UI->actionInterpolateColors,			&QAction::triggered, this, &MainWindow::doActionInterpolateColors);
 	connect(m_UI->actionEnhanceRGBWithIntensities,	&QAction::triggered, this, &MainWindow::doActionEnhanceRGBWithIntensities);
+	connect(m_UI->actionColorFromScalarField,       &QAction::triggered, this, &MainWindow::doActionColorFromScalars);
 	connect(m_UI->actionClearColor, &QAction::triggered, this, [=]() {
 		clearSelectedEntitiesProperty( ccEntityAction::CLEAR_PROPERTY::COLORS );
 	});
 
+
```
Comment (asmaloney) [inline qCC/mainwindow.cpp]: Remove blank line

File: qCC/ccColorFromScalarDlg.cpp (hunk 18 above)
Comment (samthiele) [reply qCC/ccColorFromScalarDlg.cpp]: Probably - I just couldn't figure out how to construct ccColor::Rgba instances directly from integer values (r,g,b,a). I'm sure this is me being ignorant though.... Perhaps @dgirardeau can enlighten me!

File: qCC/ccColorFromScalarDlg.cpp [hunk 25]
```diff
@@ -459,18 +452,20 @@ void ccColorFromScalarDlg::onApply()
 					col[i] = 1.0f - col[i];
 				}
 			}
-
+				
 			//calculate and set colour
 			ccColor::Rgb rgb = ccColor::Convert::hsv2rgb(col[0] * 360.0f, col[1], col[2]);
-			m_cloud->setPointColor(p, ccColor::FromQColor(QColor(int(rgb.r), int(rgb.g), int(rgb.b), int(col[3] * 255))));
+			m_cloud->setPointColor(p, ccColor::FromQColor(QColor(
```
Comment (dgirardeau) [inline qCC/ccColorFromScalarDlg.cpp]: You can directly use ccColor::Rgba(rgb, static_cast<ColorCompType>(col[3] * ccColor::MAX))

File: qCC/ccColorFromScalarDlg.cpp [hunk 26]
```diff
@@ -221,7 +217,10 @@ void ccColorFromScalarDlg::updateColormaps()
 		m_colors[0]->clear();
 
 		ccColor::Rgb col = ccColor::Convert::hsv2rgb(360.0f, 1.0f, 1.0f);
-		m_colors[0]->insert(ccColorScaleElement(1.0, QColor(int(col.r), int(col.g), int(col.b), 255)));
+		m_colors[0]->insert(ccColorScaleElement(1.0, QColor(
```
Comment (dgirardeau) [inline qCC/ccColorFromScalarDlg.cpp]: You can directly use ccColor::Rgba(col, ccColor::MAX)

File: qCC/ccColorFromScalarDlg.cpp [hunk 27]
```diff
@@ -236,7 +235,10 @@ void ccColorFromScalarDlg::updateColormaps()
 			col = ccColor::Convert::hsv2rgb(hue, 1.0f, 1.0f);
 
 			//add stop
-			m_colors[0]->insert(ccColorScaleElement( double(i) / 360.0, QColor( int(col.r), int(col.g), int(col.b), 255)));
+			m_colors[0]->insert(ccColorScaleElement( static_cast<double>(i) / 360.0, QColor(
```
Comment (dgirardeau) [inline qCC/ccColorFromScalarDlg.cpp]: You can directly use ccColor::Rgba(col, ccColor::MAX)

File: qCC/ccColorFromScalarDlg.h [hunk 28]
```diff
@@ -0,0 +1,107 @@
...
+	//saturation values
+	double m_minSat[4];
+	double m_maxSat[4];
+	//! Associated colour scales
+	ccColorScale::Shared m_colors[4];
+	//! Associated point cloud (color source)
+	ccPointCloud* m_cloud;
+
+private:
+	Ui::ColorFromScalarDialog *ui;
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.h]: It's a member variable, so generally we use `m_ui` for these.

I apologize - I should have mentioned this in my earlier comment...

File: qCC/ccColorFromScalarDlg.h [hunk 29]
```diff
@@ -0,0 +1,107 @@
...
+	void toggleColorMode(bool state);
+	void minChanged(int n, double val, bool slider);
+	void maxChanged(int n, double val, bool slider);
+	//done!
+	void onApply();
+
+protected:
+	//! Associated histogram view
+	ccHistogramWindow *m_histograms[4];  //0 - red, 1 - green, 2 - blue
+	 //scalar fields
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.h]: Remove extra space before comment?

File: qCC/ccColorFromScalarDlg.cpp [hunk 30]
```diff
@@ -0,0 +1,471 @@
...
+	connect(m_histograms[0], &ccHistogramWindow::sfMinSatValChanged, this, &ccColorFromScalarDlg::minChangedR);
+	connect(m_histograms[0], &ccHistogramWindow::sfMaxSatValChanged, this, &ccColorFromScalarDlg::maxChangedR);
+	connect(m_histograms[1], &ccHistogramWindow::sfMinSatValChanged, this, &ccColorFromScalarDlg::minChangedG);
+	connect(m_histograms[1], &ccHistogramWindow::sfMaxSatValChanged, this, &ccColorFromScalarDlg::maxChangedG);
+	connect(m_histograms[2], &ccHistogramWindow::sfMinSatValChanged, this, &ccColorFromScalarDlg::minChangedB);
+	connect(m_histograms[2], &ccHistogramWindow::sfMaxSatValChanged, this, &ccColorFromScalarDlg::maxChangedB);
+	connect(m_histograms[3], &ccHistogramWindow::sfMinSatValChanged, this, &ccColorFromScalarDlg::minChangedA);
+	connect(m_histograms[3], &ccHistogramWindow::sfMaxSatValChanged, this, &ccColorFromScalarDlg::maxChangedA);
+
+	//initialise colour scales
```
Comment (asmaloney) [inline qCC/ccColorFromScalarDlg.cpp]: +1 for proper spelling of `colour` :-)

File: qCC/ccColorFromScalarDlg.cpp (hunk 26 above)
Comment (samthiele) [reply qCC/ccColorFromScalarDlg.cpp]: ccColorScaleElement doesn't seem to have a constructor for (double, ccColor::Rgba) - the only one I can find is for (double, QColor)?

Comment (dgirardeau) [reply qCC/ccColorFromScalarDlg.cpp]: Oops sorry, I thought it was the same as the others. At least you don't need to explicitly cast to 'int' the values here.

**Pull request merged**